### PR TITLE
[trainer] feat: Add Nemo-Automodel as alternative training engine

### PR DESCRIPTION
**Pull request by HuiyingLi**
### What does this PR do?

Add NeMo-Automodel as a training engine. The SFT trainer is tested with Qwen2.5-0.5B. 
- automodel engine matches exactly with FSDP engine for SFT trainer (TP1/TP2/rmpad=True/False)
- use_remove_padding=True matches use_remove_padding=False
- EP support tested with kimi moonlight 16B

Relevant PRs:
- RFC https://github.com/verl-project/verl/issues/5245
- Add VeOmni #4072 
- Add TorchTitan #5051 

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

automodel backend 1gpu & 4gpu tp1/tp2 against FSDP backend 1gpu, rmpad true/false.
image

automodel backend finetuning moonlight 16B with ep8 8H100
image

automodel backend finetuning Qwen3 30B with ep8 8H100
image

automodel backend finetuning qwen2.5-7b with 4H100 fsdp2
image

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/workers/engine/automodel/__init__.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2025 Bytedance Ltd. and/or its affiliates
```
Comment (ISEEKYAN) [inline verl/workers/engine/automodel/__init__.py]: it is ok to mark copyright of NVIDIA, same below

**Comment by ISEEKYAN** [comment]
hi @HuiyingLi , thanks to your great contribution. 
I found the MFU of automodel is lower than FSDP on 0.5B model and the MFU is less than 1% on 16B MoE model, is this supposed to be right? Could you provide some fair comparison on popular models such as 7B dense or 30B MoE?

**Comment by HuiyingLi** [comment]
> hi @HuiyingLi , thanks to your great contribution. I found the MFU of automodel is lower than FSDP on 0.5B model and the MFU is less than 1% on 16B MoE model, is this supposed to be right? Could you provide some fair comparison on popular models such as 7B dense or 30B MoE?

Hi @ISEEKYAN , 
Thank you!
- for 0.5B model, FSDP engine is single GPU while automodel is 4 GPU. I've updated the chart with automodel single GPU for comparison.
- for the low mfu on 16B model: due to very small seqlen and batch size. I've updated the chart with larger seqlen and gbs, and added charts for Qwen 30B MoE and Qwen 7B dense.

**Comment by ISEEKYAN** [comment]
great, is your exp on H100? If true, the MFU looks good but it will be better to have a fair comparison with FSDP or Megatron. But this is not a blocker of merging this PR. It is just a good reference for the users to adopt AutoModel, and it will be better to add a doc to show the comparison and an example for the user to easily hand on

**Comment by ETOgaosion** [comment]
@HuiyingLi Thanks for your great contribution! Could you please sign the CLA~

File: docs/workers/automodel_workers.rst [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+We support the Automodel (nemo_automodel) backend by implementing the
+``AutomodelEngine`` and ``AutomodelEngineWithLMHead`` engine classes.
+The Automodel backend delegates model building, parallelization, optimizer
+sharding, LR scheduling, gradient clipping, and checkpointing to
+nemo_automodel's infrastructure while using verl's training loop,
+data pipeline, and loss function.
+
+**Requirements**
+
+- Automodel r0.3.0
```
Comment (ETOgaosion) [inline docs/workers/automodel_workers.rst:15]: Maybe in another PR, we should refactor `docs/start/install.rst` to support all model engines and rollout engine install methods and use some displays for better choices.

**Comment by ETOgaosion** [comment]
We should also prepare some CI tests for Nemo-Automodel

**Comment by HuiyingLi** [comment]
@ETOgaosion  Thank you for reviewing! I've signed the CLA. Could you please suggest how to pass the ci for this PR? For the license error do you suggest to add Nvidia to the list? Should we add ci tests in this PR? Thank you!

**Comment by ETOgaosion** [comment]
For this Nemo-Automodel CI, wondering whether verl's docker image contains needed components of Nemo-AutoModel?

Or if there is no dependency conflicts with megatron, we can directly reuse current megatron docker images to run this backend by adding needed packages.

For Adding CI, we can use another PR to test new images, we can merge this first.

With the image, you can enable test in tests/special_e2e/sft/test_sft_engine_all.sh as you implemented here to test.

File: verl/workers/engine/automodel/__init__.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2025, NVIDIA CORPORATION. All rights reserved.
```
Comment (ETOgaosion) [inline verl/workers/engine/automodel/__init__.py:1]: License LGTM

**Comment by ETOgaosion** [comment]
@HuiyingLi For failed CIs, could you rebase main and retry?

**Review by ETOgaosion — APPROVED** [review]
